<a href="https://colab.research.google.com/github/frederikvanduuren/Project-02-TPOT/blob/master/tpot_gen_pop_digits.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## The idea is to search for the sweet spot between looping generations vs populationsize.
## Will put the results in a sheet & visualize the potential sweet spot
## Will apply the same process onto different dataset in future posts
## Also keep track of Best Model!

NumPy, SciPy, scikit-learn, pandas and joblib can be installed in Anaconda via the command:

In [0]:
pip install numpy scipy scikit-learn pandas joblib

Optionally, you can install XGBoost if you would like TPOT to use the eXtreme Gradient Boosting models. XGBoost is entirely optional, and TPOT will still function normally without XGBoost if you do not have it installed

In [0]:
pip install xgboost

Finally to install TPOT itself, run the following command:

In [0]:
pip install tpot

Import TPOTClassifier

In [0]:
from tpot import TPOTClassifier

Load and return the digits dataset (classification).

In [0]:
from sklearn.datasets import load_digits

Split datasets into random train and test subsets

In [0]:
from sklearn.model_selection import train_test_split

In [0]:

import time

digits = load_digits() #load the digit dataset

# Split datasets into random train and test subsets
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target,
                                                    train_size=0.80, test_size=0.20)

gen = [5, 10, 15, 20, 25, 30, 50, 100]
pop = [5, 10, 15, 20, 25, 30, 50, 100]

for gens in gen:
  for pops in pop:
    tpot = TPOTClassifier(generations=gens, population_size=pops, verbosity=2)
    start = time.time()
    print('#generations = ', gens, '#population = ', pops)
    tpot.fit(X_train, y_train)
    end = time.time()
    print('time: ',(end-start))
    print('accuracy :', tpot.score(X_test, y_test))
    tpot.export('tpot_mnist_like_digits_gens{}_pops{}_{}.py'.format(gens,pops,tpot.score(X_test, y_test)))


#generations =  5 #population =  5


Generation 1 - Current best internal CV score: 0.9658960216041091
Generation 2 - Current best internal CV score: 0.9658960216041091
Generation 3 - Current best internal CV score: 0.9658960216041091
Generation 4 - Current best internal CV score: 0.9673144085609003
Generation 5 - Current best internal CV score: 0.9673144085609003

Best pipeline: ExtraTreesClassifier(input_matrix, bootstrap=True, criterion=entropy, max_features=0.6500000000000001, min_samples_leaf=1, min_samples_split=14, n_estimators=100)
time:  155.80346179008484
accuracy : 0.9666666666666667
#generations =  5 #population =  10


Generation 1 - Current best internal CV score: 0.9506553541453504
Generation 2 - Current best internal CV score: 0.9701095328176429
Generation 3 - Current best internal CV score: 0.9715056958445414
Generation 4 - Current best internal CV score: 0.9791307926582581
Generation 5 - Current best internal CV score: 0.9791307926582581

Best pipeline: KNeighborsClassifier(PCA(input_matrix, iterated_power=9, svd_solver=randomized), n_neighbors=9, p=1, weights=uniform)
time:  720.112889289856
accuracy : 0.9861111111111112
#generations =  5 #population =  15


Generation 1 - Current best internal CV score: 0.9770179778288979
Generation 2 - Current best internal CV score: 0.9770179778288979
Generation 3 - Current best internal CV score: 0.9798033892221669
Generation 4 - Current best internal CV score: 0.9798033892221669
Generation 5 - Current best internal CV score: 0.9819112446565732

Best pipeline: KNeighborsClassifier(SelectFwe(ExtraTreesClassifier(input_matrix, bootstrap=False, criterion=entropy, max_features=0.2, min_samples_leaf=13, min_samples_split=20, n_estimators=100), alpha=0.03), n_neighbors=1, p=1, weights=distance)
time:  428.8748104572296
accuracy : 0.975


/usr/local/lib/python3.6/dist-packages/sklearn/feature_selection/univariate_selection.py:721: RuntimeWarning: invalid value encountered in less
  return (self.pvalues_ < self.alpha / len(self.pvalues_))
/usr/local/lib/python3.6/dist-packages/sklearn/feature_selection/univariate_selection.py:721: RuntimeWarning: invalid value encountered in less
  return (self.pvalues_ < self.alpha / len(self.pvalues_))


#generations =  5 #population =  20


Generation 1 - Current best internal CV score: 0.9687177114063961
Generation 2 - Current best internal CV score: 0.9707964282503818
Generation 3 - Current best internal CV score: 0.9728655804624218
Generation 4 - Current best internal CV score: 0.9728655804624218
Generation 5 - Current best internal CV score: 0.975628767925798

Best pipeline: GradientBoostingClassifier(input_matrix, learning_rate=0.1, max_depth=3, max_features=0.15000000000000002, min_samples_leaf=18, min_samples_split=15, n_estimators=100, subsample=0.35000000000000003)
time:  448.85129952430725
accuracy : 0.9777777777777777
#generations =  5 #population =  25


Generation 1 - Current best internal CV score: 0.9860755737734372
Generation 2 - Current best internal CV score: 0.9860755737734372
Generation 3 - Current best internal CV score: 0.9860755737734372
Generation 4 - Current best internal CV score: 0.986794386767907
Generation 5 - Current best internal CV score: 0.986794386767907

Best pipeline: LogisticRegression(PolynomialFeatures(MaxAbsScaler(input_matrix), degree=2, include_bias=False, interaction_only=False), C=5.0, dual=False, penalty=l2)
time:  1594.4314622879028
accuracy : 0.9861111111111112
#generations =  5 #population =  30


Generation 1 - Current best internal CV score: 0.977714978746343
Generation 2 - Current best internal CV score: 0.9826397705145167
Generation 3 - Current best internal CV score: 0.9826397705145167
Generation 4 - Current best internal CV score: 0.9826397705145167
Generation 5 - Current best internal CV score: 0.9826397705145167

Best pipeline: KNeighborsClassifier(input_matrix, n_neighbors=2, p=2, weights=distance)
time:  988.7527179718018
accuracy : 0.9777777777777777
#generations =  5 #population =  50


# Conclusion
## The Sweet Spot seems to be where Gen*Pop <= 400
## Most Common Classifier is 		KNeighborsClassifier

Generations								
50	97.78%	98.89%	98.61%	98.89%	98.33%	98.33%	98.33%	
30	98.61%	98.89%	99.44%	99.17%	98.89%	98.61%	99.17%	
25	98.61%	99.77%	99.17%	98.61%	98.61%	99.44%	99.17%	
20	97.78%	98.06%	98.61%	98.89%	98.89%	98.61%	98.89%	
15	97.50%	98.06%	98.61%	99.44%	98.89%	98.61%	98.33%	
10	98.33%	98.61%	97.78%	98.33%	98.61%	99.17%	98.89%	
5	  96.39%	95.28%	97.78%	98.61%	98.89%	98.89%	97.78%	
	  5	      10	    15	    20	    25	    30	    50	   Population

Generations								
50	KNeighborsClassifier	KNeighborsClassifier	KNeighborsClassifier	ExtraTreesClassifier	ExtraTreesClassifier	ExtraTreesClassifier	KNeighborsClassifier	
30	LogisticRegression	LogisticRegression	LinearSVC	LogisticRegression	ExtraTreesClassifier	KNeighborsClassifier	KNeighborsClassifier	
25	KNeighborsClassifier	KNeighborsClassifier	GradientBoostingClassifier	LogisticRegression	LogisticRegression	KNeighborsClassifier	LogisticRegression	
20	KNeighborsClassifier	KNeighborsClassifier	KNeighborsClassifier	LinearSVC	KNeighborsClassifier	KNeighborsClassifier	ExtraTreesClassifier	
15	KNeighborsClassifier	LogisticRegression	LogisticRegression	LinearSVC	KNeighborsClassifier	KNeighborsClassifier	LinearSVC	
10	LinearSVC	KNeighborsClassifier	ExtraTreesClassifier	ExtraTreesClassifier	ExtraTreesClassifier	LogisticRegression	KNeighborsClassifier	
5	LogisticRegression	LogisticRegression	ExtraTreesClassifier	LogisticRegression	KNeighborsClassifier	KNeighborsClassifier	KNeighborsClassifier	
	5	10	15	20	25	30	50	Population

## Count
### LinearSVC	5
## KNeighborsClassifier	22
## LogisticRegression	12
## GradientBoostingClassifier	1
## ExtraTreesClassifier	9